### PR TITLE
feat(meshexternalservice): initial docs

### PR DESCRIPTION
**Pull request by slonka**
closes https://github.com/kumahq/kuma-website/issues/1815

Did you sign your commit? [Instructions](https://github.com/kumahq/.github/blob/main/CONTRIBUTING.md#sign-your-commits)

Have you read [Contributing guidelines](https://github.com/kumahq/.github/blob/main/CONTRIBUTING.md)?

rendered https://deploy-preview-1816--kuma.netlify.app/docs/dev/networking/meshexternalservice/

### CODE REVIEW

File: app/_src/policies/meshexternalservice.md [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+### DNS setup
+
+To be able to access `MeshExternalService` via a hostname you need to define a [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) with a `meshExternalService` selector.
+
+## Examples
+
+For the following examples the following `HostnameGenerator` will be used:
+
+{% policy_yaml hostnamegenerator %}
```
Comment (lukidzi) [inline app/_src/policies/meshexternalservice.md:86]: I think we should add explanations to examples

File: app/_src/policies/meshexternalservice.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        value: "spiffe://example.local/ns/local"
+    caCert:
+      inline: dGVzdA==
+    clientCert:
+      secret: "123"
+    clientKey:
+      secret: "123"
+```
+
+When TLS is enabled but the CA is not set, the sidecar uses the default OS-specific CA. The user can override the default CA by setting the path in the environment variable `KUMA_DATAPLANE_RUNTIME_DYNAMIC_SYSTEM_CA_PATH` for the sidecar
```
Comment (slonka) [inline app/_src/policies/meshexternalservice.md:76]:
```suggestion
When TLS is enabled but `caCert` is not set, the sidecar uses the [autodetected OS-specific CA](https://github.com/kumahq/kuma/blob/aba6518fca65bc7ab52e5328eb686a51a6f98a53/app/kuma-dp/pkg/dataplane/certificate/cert.go#L12). The user can override the default CA by setting the path in the environment variable `KUMA_DATAPLANE_RUNTIME_DYNAMIC_SYSTEM_CA_PATH` for the sidecar.
```

File: app/_src/policies/meshexternalservice.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+When TLS is enabled but `caCert` is not set, the sidecar uses the [autodetected OS-specific CA](https://github.com/kumahq/kuma/blob/aba6518fca65bc7ab52e5328eb686a51a6f98a53/app/kuma-dp/pkg/dataplane/certificate/cert.go#L12). The user can override the default CA by setting the path in the environment variable `KUMA_DATAPLANE_RUNTIME_DYNAMIC_SYSTEM_CA_PATH` for the sidecar.
+
+### DNS setup
+
+To be able to access `MeshExternalService` via a hostname you need to define a [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) with a `meshExternalService` selector.
+
+## Examples
+
+For the following examples the following `HostnameGenerator` will be used:
```
Comment (slonka) [inline app/_src/policies/meshexternalservice.md:84]:
```suggestion
For the examples below the following `HostnameGenerator` will be used:
```

File: app/_src/policies/meshexternalservice.md (hunk 1 above)
Comment (slonka) [reply app/_src/policies/meshexternalservice.md:90]: added

File: app/_src/policies/meshexternalservice.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+      inline: dGVzdA==
+    clientCert:
+      secret: "123"
+    clientKey:
+      secret: "123"
+```
+
+When TLS is enabled but `caCert` is not set, the sidecar uses the [autodetected OS-specific CA](https://github.com/kumahq/kuma/blob/aba6518fca65bc7ab52e5328eb686a51a6f98a53/app/kuma-dp/pkg/dataplane/certificate/cert.go#L12). The user can override the default CA by setting the path in the environment variable `KUMA_DATAPLANE_RUNTIME_DYNAMIC_SYSTEM_CA_PATH` for the sidecar.
+
+### DNS setup
```
Comment (lukidzi) [inline app/_src/policies/meshexternalservice.md:78]: We could add some information about rich clients (http/tcp clients: like kafka) and link passthrough docs somewhere.

File: app/_src/policies/external-services.md [hunk 5]
```diff
@@ -1,6 +1,9 @@
 ---
 title: External Service
 ---
+{% if_version gte:2.8.x %}
+New to Kuma? You don't need this, check [`MeshExternalService`](/docs/{{ page.version }}/policies/meshexternalservice) instead.
```
Comment (jakubdyszkiewicz) [inline app/_src/policies/external-services.md:5]: so, is it going GA? Does policy matching work?

File: app/_src/policies/meshexternalservice.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
+---
+title: MeshExternalService
+---
+
+{% warning %}
+Do **not** combine with [External Service](/docs/{{ page.version }}/policies/external-services).
```
Comment (jakubdyszkiewicz) [inline app/_src/policies/meshexternalservice.md:6]: what happens then? How should I migrate from ExternalService then if I can't combine it?

File: app/_src/policies/meshexternalservice.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+{% endwarning %}
+
+This resource allows services running inside the mesh to consume services that are not part of the mesh.
+The `MeshExternalService` resource allows you to declare external resources instead of relying on [MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough) or [passthrough mode](/docs/{{ page.version }}/networking/non-mesh-traffic#outgoing).
+In contrast to passthrough `MeshExternalService` behaves like a normal service as if it was part of the mesh.
```
Comment (jijiechen) [inline app/_src/policies/meshexternalservice.md:11]:
```suggestion
In contrast to passthrough, `MeshExternalService` behaves like a normal service as if it is part of the mesh.
```

File: app/_src/policies/meshexternalservice.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+{% warning %}
+Currently `MeshExternalService` resource does not support targeting by [targetRef policies](/docs/{{ page.version }}/policies/targetref).
+This limitation will be lifted in the next release.
```
Comment (jijiechen) [inline app/_src/policies/meshexternalservice.md:15]:
```suggestion
This limitation will be lifted in a future release.
```

File: app/_src/policies/meshexternalservice.md (hunk 7 above)
Comment (jakubdyszkiewicz) [inline app/_src/policies/meshexternalservice.md:11]: as a first-time reader it's still not clear what's the difference and what it means that it's "normal service".

I would try to rephrase it so that mesh passthrough simply allows traffic to flow while MeshExternalService allows us to use policies like MeshTrafficPermission, MeshRateLimit etc.

File: app/_src/policies/meshexternalservice.md (hunk 6 above)
Comment (slonka) [reply app/_src/policies/meshexternalservice.md:6]: I need to check what happens if you create a hostname generator that overshadows a domain hijacked by regular ExternalService.

File: app/_src/policies/external-services.md (hunk 5 above)
Comment (slonka) [reply app/_src/policies/external-services.md:5]: AFAIK thanks to @lukidzi policy matching works if you're using egress (so the same way it used to work so people are not losing any features). I'll do examples of this.

File: app/_src/policies/meshexternalservice.md [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
...CnBsU3FyWFRaZCtLRlhTdzBwNDNwMmltN2hzSmxmNUlHN3AwYWFEdUJlZUdsVmhGT3E5dU5CRmgxMWN4ajgza2gKZVI1Q1RWZXpjV1NhWExLN3pMUjdqNUFjdTlyaU1tWGFEUUdJaWQrNU83cGFnRFNtNHpOSENZdnFKYm1XcU5DWApOcXVsNFI3L3NoUGd3MjQ0R0J6TmxSdjJZUUtCZ1FEQzhrWGM4UWt1amtwS0pkVm8wWFZFd1MzdWFmb3lVeGNxClBmb2dWOEpsUE1GRCs5VjJjNkhSMkpJWktKMzU0RFhsby8ydWllZzhQT0lUMXBpVlVCQk55eElQY1FxMUcweFcKQmtPMkpyMmdyZlN6KzNwM2RzTGR5dXU5bWc4eVBiNWhwRVdjTjFtUDFiWko3YUs0Qjk5SFArczR2TXd1c2xWOApzZDI2Uk1YV2RRS0JnRnRqeWhkVGVKU1poY25GbXdYQk9BMlJGQmN2UER3Q3NlOFpGY0dHcmU1VWxYczg1aWpSCmxmNGowQjZQSkNrK1l2NlpUUTVBZVBBeHFoZlBvNDBqNWxYVnNJQWl1VDZ4NGZ1dzVuSkdvNWhEeUY1OU9qbloKbEltZ0FaREszS1hmNFhyZUl1bm93VXBSeXBlRUdEVjhBdG5nR0FaMFh3T0Z2Nm9ZZlhZVHg2ZUJBb0dBRlJQNAo5ZENoKzRTckI2VmJrNy9CL0RNZThqNUhMUlhLMVdocUdRRWtKYW9TQTNYQk9OTjcxYUtpK1ZGbzgxR0l3bEdlCjVqWkhBK3haVFdmUWk2UmlmdWJNQnh0ajJ2MGVuZGFEajdoVW5JRHlpbHRRZklZOHY1cG5MdEx2ZmJFcldvZFcKZDNPTW5YNnYvUUpTcTY4K053ZjBPT2hBODNPWXhxaThucDA4L3RrQ2dZRUFqYkNUcHhIM2hmd2p2Y2t2TFIxVQpHWVdUaVkzMWJWVDJEVTNFZjlLb09KQXJQSlRuRlNyQ0VuNEpESGk0S1NrencvRHpQRmtMRzdsSXZFV21qcTJTCnRyUGdHMEtCcWxQa1pCRkRYUy9tdWFWVzQyUFhFeVhLV1kzdFFaRnVzQk5XdVZXOFNvZW8zdTd1YWlXbHRPNzAKTnZOSW83U21MNVJ0UEF4d1NpUTJzOGc9Ci0tLS0tRU5EIFBSSVZBVEUgS0VZLS0tLS0KCg==
```
Comment (jakubdyszkiewicz) [inline app/_src/policies/meshexternalservice.md:190]: While this is ok, it would be cool to store this in Secret first. It's more realistic scenario. Up to you

Comment (slonka) [reply app/_src/policies/meshexternalservice.md:190]: Just wanted to make this as easy as possible. I can add a note that in a real world scenario you want to use `secret` and store it as a k8s secret instead of inline, WDYT?

**Review by jakubdyszkiewicz — COMMENTED** [review]
very cool examples. I really like this 👍

File: app/_src/policies/meshexternalservice.md (hunk 9 above)
Comment (jakubdyszkiewicz) [reply app/_src/policies/meshexternalservice.md:190]: I think that works as well. Also link to Kuma secrets page.

File: app/_src/policies/external-services.md (hunk 5 above)
Comment (lukidzi) [reply app/_src/policies/external-services.md:5]: It works for MeshTrafficPermisison

Comment (slonka) [reply app/_src/policies/external-services.md:5]: Correction: It's only MTP so it can't be here.

File: app/_src/policies/external-services.md [hunk 10]
```diff
@@ -1,6 +1,9 @@
...
+{% endif_version %}
```
Comment (slonka) [inline app/_src/policies/external-services.md:6]:
```suggestion
```

File: app/_src/policies/meshexternalservice.md (hunk 4 above)
Comment (slonka) [reply app/_src/policies/meshexternalservice.md:78]: I've added info + link, do you think it's enough?

File: app/_src/policies/external-services.md (hunk 5 above)
Comment (slonka) [reply app/_src/policies/external-services.md]: Added a note that the resource is experimental (the same way MeshService is).

File: app/_src/policies/meshexternalservice.md (hunk 6 above)
Comment (slonka) [reply app/_src/policies/meshexternalservice.md]: Added a note in DNS setup section to not hijack real DNS addresses.

File: app/_src/policies/meshexternalservice.md (hunk 9 above)
Comment (slonka) [reply app/_src/policies/meshexternalservice.md:201]: added a note above

File: app/_src/policies/meshexternalservice.md (hunk 7 above)
Comment (slonka) [reply app/_src/policies/meshexternalservice.md]: I removed that sentence because targeting does not work yet (I thought on egress it did) and added a note that the resource is experimental.

File: app/_data/docs_nav_kuma_dev.yml [hunk 11]
```diff
@@ -287,6 +287,13 @@ items:
             url: "/policies/meshcircuitbreaker/#configuration"
           - text: Examples
             url: "/policies/meshcircuitbreaker/#examples"
+      - text: MeshExternalService
```
[the remark this answers is not in the document]
Comment (slonka) [reply app/_data/docs_nav_kuma_dev.yml]: this is a suggestion from @lahabana - FYI @michaelbeaumont

File: app/_src/networking/meshexternalservice.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+Currently `MeshExternalService` resource only supports targeting by [MeshTrafficPermission](/docs/{{ page.version }}/policies/meshtrafficpermission) with [Zone Egress](/docs/{{ page.version }}/production/cp-deployment/zoneegress).
+This limitation will be lifted in the next release.
+{% endwarning %}
+
+## Configuration
+
+### Match
+
+This section specifies the rules for matching traffic that will be routed to external resources defined in `endpoints` section.
+The only `type` supported is [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) and it means that it will match traffic directed to a hostname created by the hostname generator.
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:22]: hostnamegenerator is not a policy

Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:22]: If there's only one supported type should the type be optional?

File: app/_src/networking/meshexternalservice.md [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+When TLS is enabled but `caCert` is not set, the sidecar uses the [autodetected OS-specific CA](https://github.com/kumahq/kuma/blob/aba6518fca65bc7ab52e5328eb686a51a6f98a53/app/kuma-dp/pkg/dataplane/certificate/cert.go#L12).
+The user can override the default CA by setting the path in the environment variable `KUMA_DATAPLANE_RUNTIME_DYNAMIC_SYSTEM_CA_PATH` for the sidecar.
+
+### DNS setup
+
+To be able to access `MeshExternalService` via a hostname you need to define a [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) with a `meshExternalService` selector.
+
+{% warning %}
+Do **not** hijack original addresses like httpbin.com (the way it was done with [External Service](/docs/{{ page.version }}/policies/external-services)).
+If you need to transparently pass traffic through the Mesh without modifying it use [MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough).
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:84]: Might be worth explaining why.

File: app/_src/networking/meshexternalservice.md [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+For clients that the hostname is not fully known in advance please take a look at [Wildcard DNS matching in MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough/#wildcard-dns-matching).
+
+## Examples
+
+TCP examples use https://tcpbin.com/ service which is a TCP echo service, check out the website for more details.
+HTTP examples use https://httpbin.org/ service which is a website for inspecting and debugging HTTP requests.
+GRPC examples use https://grpcbin.test.k6.io/ service which is a gRPC Request & Response Service.
+You can use [grpcurl](https://github.com/fullstorydev/grpcurl) as a client, it is available in [netshoot](https://github.com/nicolaka/netshoot) debug image alongside other tools used in later sections.
+
+For the examples below the following `HostnameGenerator` will be used:
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:96]: So there's no default hostname generator atm?

File: app/_src/networking/meshexternalservice.md [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+spec:
+  selector:
+    meshExternalService:
+      matchLabels: {}
+  template: '{{ .Name }}.mesh'
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:113]: I'd stay away from this as an example as this may collide with default hostnames in older Kuma.

File: app/_src/networking/meshexternalservice.md (hunk 14 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:96]: I think te answer is here https://github.com/kumahq/kuma/pull/10478

File: app/_src/networking/meshexternalservice.md [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+### DNS setup
+
+To be able to access `MeshExternalService` via a hostname you need to define a [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) with a `meshExternalService` selector.
+
+{% warning %}
+Do **not** hijack original addresses like httpbin.com (the way it was done with [External Service](/docs/{{ page.version }}/policies/external-services)).
+If you need to transparently pass traffic through the Mesh without modifying it use [MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough).
+{% endwarning %}
+
+For clients that the hostname is not fully known in advance please take a look at [Wildcard DNS matching in MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough/#wildcard-dns-matching).
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:87]: Might be worth adding a few lines after this to add info about what this is doing internally:

- one or multiple hostname are generated by using matching hostname generators
- vip allocated
- envoy service created which will use `endpoints` as the different endpoints

File: app/_src/networking/meshexternalservice.md (hunk 14 above)
Comment (lahabana) [reply app/_src/networking/meshexternalservice.md:96]: Might be worth stating this in the docs that in future versions of kuma there will be a default HostnameGenerator

Comment (lahabana) [reply app/_src/networking/meshexternalservice.md:96]: And write the docs in a way that will be easy to make this part optional

File: app/_src/networking/meshexternalservice.md (hunk 12 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:22]: we're just pointing to whatever is here https://github.com/kumahq/kuma-website/pull/1814#discussion_r1639391091

Comment (slonka) [reply app/_src/networking/meshexternalservice.md:22]: yes, let's make it optional

File: app/_src/networking/meshexternalservice.md (hunk 14 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:96]: > And write the docs in a way that will be easy to make this part optional

So you'd use the template that is suggested currently so we don't have to if/else the examples? Otherwise I'd just if/else https://github.com/kumahq/kuma-website/pull/1816/files#diff-8d43902ad147c9062f03d2eea50710e9ba83961b78556a27086f0e8b0b8a0d05R98-R131 this part.

File: app/_src/networking/meshexternalservice.md [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+{% endtabs %}"
+
+### TCP
+
+This is a simple example of accessing `tcpbin.com` service without TLS that echos back bytes sent to it.
+
+{% policy_yaml tcp %}
+```yaml
+apiVersion: kuma.io/v1alpha1
+type: MeshExternalService
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:140]: Your examples all add an `apiVersion` because you mix up k8s and universal format here

File: app/_src/networking/meshexternalservice.md (hunk 12 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:22]: https://github.com/kumahq/kuma/pull/10570/files

File: app/_src/networking/meshexternalservice.md (hunk 13 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:92]: added a note about hijacking original address - let me know if that is what you mean or please suggest something

File: app/_src/networking/meshexternalservice.md (hunk 12 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:22]: done

**Review by jakubdyszkiewicz — APPROVED** [review]

File: app/_src/networking/meshexternalservice.md [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+- a VIP is allocated from `242.0.0.0/8` range (can be changed by `KUMA_IPAM_MESH_EXTERNAL_SERVICE_CIDR` environment variable)
+- Envoy cluster is created which will use endpoints defined in `spec.endpoints` as the [cluster endpoints](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto)
+
+{% warning %}
+Do **not** hijack original addresses like httpbin.com (the way it was done with [External Service](/docs/{{ page.version }}/policies/external-services)).
+Hijacking the original address is like performing a man-in-the-middle attack so there is a high chance of something breaking.
+If you need to transparently pass traffic through the Mesh without modifying it use [MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough).
+{% endwarning %}
+
+For clients that the hostname is not fully known in advance please take a look at [Wildcard DNS matching in MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough/#wildcard-dns-matching).
```
Comment (lahabana) [inline app/_src/networking/meshexternalservice.md:95]:
```suggestion
For accessing entire sets of subdomains, take a look at [Wildcard DNS matching in MeshPassthrough](/docs/{{ page.version }}/policies/meshpassthrough/#wildcard-dns-matching).
```

File: jekyll-kuma-plugins/lib/jekyll/kuma-plugins/liquid/tags/policyyaml.rb [hunk 19]
```diff
@@ -52,16 +57,18 @@ def process_array(array, remove_suffix, rename_suffix)
 
           def render(context)
             content = super
-            return "" unless content != ""
+            return "" if content == ""
+            has_raw = @body.nodelist.first { |x| x.has?("tag_name") and x.tag_name == "raw"}
```
Comment (michaelbeaumont) [inline jekyll-kuma-plugins/lib/jekyll/kuma-plugins/liquid/tags/policyyaml.rb:61]: Please make this in a separate PR next time :pray:

Comment (slonka) [reply jekyll-kuma-plugins/lib/jekyll/kuma-plugins/liquid/tags/policyyaml.rb:61]: I can make it a separate PR if you want.

File: app/_src/networking/meshexternalservice.md (hunk 12 above)
Comment (slonka) [reply app/_src/networking/meshexternalservice.md:22]:
```suggestion
The only `type` supported is `HostnameGenerator` (this field is optional so can be omitted) and it means that it will match traffic directed to a hostname created by the hostname generator. 
```

File: app/_src/networking/meshexternalservice.md [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+
+When TLS is enabled but `caCert` is not set, the sidecar uses the [autodetected OS-specific CA](https://github.com/kumahq/kuma/blob/aba6518fca65bc7ab52e5328eb686a51a6f98a53/app/kuma-dp/pkg/dataplane/certificate/cert.go#L12).
+The user can override the default CA by setting the path in the environment variable `KUMA_DATAPLANE_RUNTIME_DYNAMIC_SYSTEM_CA_PATH` for the sidecar.
+
+### DNS setup
+
+To be able to access `MeshExternalService` via a hostname you need to define a [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) with a `meshExternalService` selector.
+In the future release a default [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) will be provided.
+
+Once a [HostnameGenerator](/docs/{{ page.version }}/policies/hostnamegenerator) and a `MeshExternalService` is in place the following will happen:
```
Comment (slonka) [inline app/_src/networking/meshexternalservice.md:84]:
```suggestion
To be able to access `MeshExternalService` via a hostname you need to define a `HostnameGenerator` with a `meshExternalService` selector.
In the future release a default `HostnameGenerator` will be provided.

Once a `HostnameGenerator` and a `MeshExternalService` is in place the following will happen:
```
